## LogisticRegression model

- Logistic regression

Logistic regression is named for the function used at the core of the method, the logistic function.
also called the sigmoid function was developed by statisticians to describe properties of population growth in ecology, rising quickly and maxing out at the carrying capacity of the environment. It’s an S-shaped curve that can take any real-valued number and map it into a value between 0 and 1.

1 / (1 + e^-value)

- load dataset

In [1]:
# import library
import matplotlib.pyplot as plt
from sklearn.preprocessing import PolynomialFeatures
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from subprocess import check_output
from datetime import time
from sklearn.metrics import classification_report
from sklearn.linear_model import LogisticRegression
import joblib

# load data
data = pd.read_csv("../../data.csv")

# Get valuse from data (datafram)
data = data.values
X = data[:,0:6] # all rows, no Lebal
y = data[:,6] # all rows, label only

# expand dataset features
# Define PolynomialFeatures variable with degree 5 and without bias
poly = PolynomialFeatures(degree=5 , include_bias=False) # Define PolynomialFeatures variable
X=poly.fit_transform(X) # fit PolynomialFeatures
X.shape # print data shape after PolynomialFeatures 


(11077, 461)

- split dataset to train and test dataset
    - with ratio 75% to train the dataset and 25% to the test dataset

In [2]:
# split dataset to train and test dataset 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)

#### Train LogisticRegression Model
- Define the LogisticRegression model
- Train(fit) the LogisticRegression model
- Predict the output of X_test using LogisticRegression model
- Calculate the accuracy of LogisticRegression model

In [3]:
logreg = LogisticRegression();
logreg.fit(X_train,y_train)
y_predict = logreg.predict(X_test)
score = logreg.score(X_test, y_test)

print (score)

0.5660649819494585


c:\users\hoomo\anaconda3\envs\tf2\lib\site-packages\sklearn\linear_model\_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


In [4]:
y_predict[0]

1.0

#### Evaluate Model 
- clasification report 
- print model accuracy (score = 0.5661)

In [5]:
# clasification report 
print(classification_report(y_test, y_predict))

              precision    recall  f1-score   support

         1.0       0.81      0.81      0.81       498
         2.0       0.77      0.76      0.77       480
         3.0       0.47      0.69      0.56       654
         4.0       0.30      0.28      0.29       223
         5.0       0.51      0.33      0.40       382
         6.0       0.42      0.31      0.36       533

    accuracy                           0.57      2770
   macro avg       0.55      0.53      0.53      2770
weighted avg       0.57      0.57      0.56      2770



#### save the model after training.
- so we can load it and run without the need to train anymore

In [6]:
# save the model to disk
filename = 'LogisticRegression.sav'
joblib.dump(logreg, filename)

['LogisticRegression.sav']